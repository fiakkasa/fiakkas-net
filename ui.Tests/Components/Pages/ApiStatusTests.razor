@inherits TestContext

@code {
    private readonly ILogger<ApiStatus> _logger = Substitute.For<ILogger<ApiStatus>>();

    public ApiStatusTests()
    {
        ComponentFactories.AddStub<PageTitle>();
        Services.AddSingleton(_logger);
    }

    [Fact]
    public async Task Api_Status_Should_Render_Template_When_Data_Is_Present()
    {
        var taskSource = new TaskCompletionSource<IOperationResult<IGetSystemStatusResult>>();
        var data = new GetSystemStatusResult(
            new GetSystemStatus_SystemStatus_SystemInfoItem(
                new(2024, 1, 1, 0, 0, 0, TimeSpan.Zero),
                TimeSpan.FromMinutes(1),
                "1.0.0",
                new GetSystemStatus_SystemStatus_Health_HealthReportSummary(
                    HealthStatus.Healthy,
                    TimeSpan.FromSeconds(30),
                    [
                        new GetSystemStatus_SystemStatus_Health_Entries_KeyValuePairOfStringAndHealthReportItem(
                            "Service",
                            new GetSystemStatus_SystemStatus_Health_Entries_Value_HealthReportItem(
                                "Description",
                                TimeSpan.FromSeconds(30),
                                HealthStatus.Healthy,
                                null
                            )
                        )
                    ]
                )
            )
        );
        var result = Substitute.For<IOperationResult<IGetSystemStatusResult>>();
        result.Data.Returns(data);
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .Returns(taskSource.Task);
        Services.AddSingleton(api);

        var cut = Render(@<ApiStatus/>);

        await Task.Delay(250);

        cut.FindComponent<Skeleton>();

        taskSource.SetResult(result);

        await Task.Delay(250);

        cut.Find("table.table");

        cut.Markup.MatchSnapshot();
    }

    [Fact]
    public async Task Api_Status_Should_Render_Template_When_Data_Is_Present_With_No_Health_Entries()
    {
        var taskSource = new TaskCompletionSource<IOperationResult<IGetSystemStatusResult>>();
        var data = new GetSystemStatusResult(
            new GetSystemStatus_SystemStatus_SystemInfoItem(
                new(2024, 1, 1, 0, 0, 0, TimeSpan.Zero),
                TimeSpan.FromMinutes(1),
                "1.0.0",
                new GetSystemStatus_SystemStatus_Health_HealthReportSummary(
                    HealthStatus.Healthy,
                    TimeSpan.FromSeconds(30),
                    null
                )
            )
        );
        var result = Substitute.For<IOperationResult<IGetSystemStatusResult>>();
        result.Data.Returns(data);
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .Returns(taskSource.Task);
        Services.AddSingleton(api);

        var cut = Render(@<ApiStatus/>);

        await Task.Delay(250);

        cut.FindComponent<Skeleton>();

        taskSource.SetResult(result);

        await Task.Delay(250);

        cut.Find("table.table");

        cut.Markup.MatchSnapshot();
    }

    [Fact]
    public async Task Api_Status_Should_Render_Loading_When_Data_Is_In_Transit()
    {
        var taskSource = new TaskCompletionSource<IOperationResult<IGetSystemStatusResult>>();
        var result = Substitute.For<IOperationResult<IGetSystemStatusResult>>();
        result.Data.Returns(default(IGetSystemStatusResult?));
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .Returns(taskSource.Task);
        Services.AddSingleton(api);

        var cut = Render(@<ApiStatus/>);

        await Task.Delay(250);

        cut.FindComponent<Skeleton>();

        taskSource.SetResult(result);

        cut.Markup.MatchSnapshot();
    }

    [Fact]
    public async Task Api_Status_Should_Prevent_Refresh_When_Loading()
    {
        var taskSource = new TaskCompletionSource<IOperationResult<IGetSystemStatusResult>>();
        var result = Substitute.For<IOperationResult<IGetSystemStatusResult>>();
        result.Data.Returns(default(IGetSystemStatusResult?));
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .Returns(
                Task.FromResult(result),
                taskSource.Task
            );
        Services.AddSingleton(api);

        var cut = RenderComponent<ApiStatus>();
        var refreshButton = cut.Find("button");

        await Task.Delay(250);

        refreshButton.GetAttribute("disabled").Should().BeNull();
        cut.Instance.Loading.Should().BeFalse();

        refreshButton.Click();

        await Task.Delay(250);

        refreshButton.GetAttribute("disabled").Should().NotBeNull();
        cut.Instance.Loading.Should().BeTrue();

        refreshButton.RemoveAttribute("disabled");

        refreshButton.Click();

        await Task.Delay(250);

        cut.Instance.Loading.Should().BeTrue();

        taskSource.SetResult(result);

        cut.Markup.MatchSnapshot();
    }

    [Fact]
    public async Task Api_Status_Should_Allow_Refresh_When_Not_Loading()
    {
        var taskSourceAfter = new TaskCompletionSource<IOperationResult<IGetSystemStatusResult>>();
        var result = Substitute.For<IOperationResult<IGetSystemStatusResult>>();
        result.Data.Returns(default(IGetSystemStatusResult?));
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .Returns(
                Task.FromResult(result),
                taskSourceAfter.Task
            );
        Services.AddSingleton(api);

        var cut = RenderComponent<ApiStatus>();
        var refreshButton = cut.Find("button");

        await Task.Delay(250);

        refreshButton.GetAttribute("disabled").Should().BeNull();
        cut.Instance.Loading.Should().BeFalse();

        refreshButton.Click();

        await Task.Delay(250);

        refreshButton.GetAttribute("disabled").Should().BeEmpty();
        cut.Instance.Loading.Should().BeTrue();

        taskSourceAfter.SetResult(result);

        cut.Markup.MatchSnapshot();
    }

    [Fact]
    public async Task Api_Status_Should_Render_GenericNodata_When_Data_Is_Not_Present()
    {
        var result = Substitute.For<IOperationResult<IGetSystemStatusResult>>();
        result.Data.Returns(default(IGetSystemStatusResult?));
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .Returns(Task.FromResult(result));
        Services.AddSingleton(api);

        var cut = Render(@<ApiStatus/>);

        await Task.Delay(250);

        cut.FindComponent<GenericNoData>();

        cut.Markup.MatchSnapshot();
    }

    [Fact]
    public async Task Api_Status_Should_Render_GenericError_When_Error()
    {
        var api = Substitute.For<IFiakkasNetApi>();
        api
            .GetSystemStatus
            .ExecuteAsync(Arg.Any<CancellationToken>())
            .ThrowsAsync(new Exception("Splash!"));
        Services.AddSingleton(api);

        var cut = Render(@<ApiStatus/>);

        await Task.Delay(250);

        cut.FindComponent<GenericError>();

        cut.Markup.MatchSnapshot();
    }

}
