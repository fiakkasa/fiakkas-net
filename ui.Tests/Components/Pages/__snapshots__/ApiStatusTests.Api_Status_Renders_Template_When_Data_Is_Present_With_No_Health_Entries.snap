<h1 class="page-header sticky-top d-flex gap-2 justify-content-start align-items-center bg-body"><span>API Status</span>

        <button type="button" class="btn btn-outline-primary rounded-circle p-0 px-1" title="Refresh" blazor:onclick="1"><i class="bi bi-arrow-clockwise"></i></button></h1>

<div class="table-responsive"><table class="table table-striped table-borderless text-nowrap"><tbody><tr><th>Startup Time:</th>
                    <td class="w-100">01/01/2024 00:00:00 +02:00</td></tr>
                <tr><th>Up Time:</th>
                    <td>00:01:00</td></tr>
                <tr><th>Version:</th>
                    <td>1.0.0</td></tr>
                <tr><th>Overall Health:</th>
                    <td>Healthy</td></tr><tr><th>Total Duration:</th>
                    <td>00:00:30</td></tr></tbody></table></div>
